src/librustc_mir/build/matches/mod.rs

```diff
@@ -38,6 +38,10 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                       -> BlockAnd<()> {
         let discriminant_place = unpack!(block = self.as_place(block, discriminant));
 
+        let arms: Vec<Arm<'tcx>> = arms.into_iter().filter(|arm|
+            !arm.patterns.iter().any(|pat| pat.is_unreachable())
```

Comment:
Hmm, I'm not sure if this is quite right either. Consider a case where you have an arm with two patterns, and one of them is unreachable but the other isn't. e.g.

```rust
let x: Option<!> = None;
match x {
    Some(_) | None => ...
}
```

woudn't this (single) arm be filtered out here? Or am I confused. Maybe we want `.filter(|arm| arm.patterns.iter().any(|pat| !pat.is_unreachable()))`? i.e., keep the arm if any of its patterns are reachable? Or, probably, we want to filter out the *patterns* (and then filter out arms that have no patterns left)?

Reply:
Whoops, you're absolutely right — I put the `!` in the wrong place.